### PR TITLE
Initial implementation of requestIdleCallback on Android

### PR DESCRIPTION
**Pull request by brentvatne**
- Does not implement implement `IdleRequestOptions` or `didTimeout`
- Uses naive heuristic for finding idle time: runs after all ReactChoreographer callbacks, if there is >= 12ms remaining in frame, execute idle callbacks
- Falls back to `requestAnimationFrame` on iOS and warns that it is not implemented on this platform yet

Test plan:
- Use `requestIdleCallback` to schedule some work, then stagger CPU to be burned for ~10ms in various `requestAnimationFrame` callbacks, idle callback should not fire until all of those are completed.

ric

### CODE REVIEW

**Comment by bestander** [comment]
Hey @brentvatne note that your unit tests are crashing, check out `./gradlew :ReactAndroid:testDebugUnitTest`

**Comment by bestander** [comment]
I am working on making CI builds stable and your branch caught my attention.

**Comment by brentvatne** [comment]
Thanks @bestander - I'll check that out :)

**Comment by brentvatne** [comment]
@bestander - summary of the issue I'm having getting `TimingModuleTest` to pass follows:

- Currently the only test that isn't working is `testPausingAndResuming`
- On my machine, test execution simply freezes and does not throw any sort of error to indicate why it isn't working
- I looked into this further and found that `mTiming.onHostPause();` is the cause
- If I comment [this line](https://github.com/exponentjs/react-native/blob/2be62e4eafc0e59fab646d874299f8710af29d1e/ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java#L231-L232) out, it passes

**Comment by brentvatne** [comment]
Nevermind @bestander - I figured it out in 2c4c39af388c2a50826467ea8e13323c4be4d604 and added coverage for requestIdleCallback in 8d8da11010e478bb08c2298f9512965f1ee21037 :)

**Comment by bestander** [comment]
:+1

**Comment by astreet** [comment]
Unfortunately, this doesn't give us much of a guarantee of idleness: Choreographer (and ReactChoreographer) callbacks execute their callbacks in the CALLBACK_ANIMATION phase of the frame, before we do CALLBACK_TRAVERSALS (the standard layout/measure/draw pass): https://github.com/android/platform_frameworks_base/blob/d59921149bb5948ffbcb9a9e832e9ac1538e05a0/core/java/android/view/Choreographer.java#L600-L608. Additionally (though probably less significant), any messages that were posted to the Looper (e.g. via View#post) since the last frame's CALLBACK_TRAVERSALS was called will be executed after the current frame's CALLBACK_TRAVERSALS.

Can you give an example of what you're trying to accomplish and maybe I can give some suggestions?

File: ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java [hunk 1]
```diff
@@ -155,7 +226,9 @@ private void setChoreographerCallback() {
     if (!mFrameCallbackPosted) {
       Assertions.assertNotNull(mReactChoreographer).postFrameCallback(
           ReactChoreographer.CallbackType.TIMERS_EVENTS,
-          mFrameCallback);
+          mTimerFrameCallback);
+      Assertions.assertNotNull(mChoreographer).postFrameCallback(
```
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java]: We really don't want to use the regular Choreographer here: ReactChoreographer allows us to be explicit about ordering of callback execution within the React framework, whereas if we use Choreographer directly, we might accidentally execute this before all of the ReactChoreographer callbacks.

Comment (brentvatne) [inline ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java]: cc @kmagiera

**Comment by brentvatne** [comment]
> Unfortunately, this doesn't give us much of a guarantee of idleness: Choreographer (and ReactChoreographer) callbacks execute their callbacks in the CALLBACK_ANIMATION phase of the frame, before we do CALLBACK_TRAVERSALS (the standard layout/measure/draw pass): https://github.com/android/platform_frameworks_base/blob/d59921149bb5948ffbcb9a9e832e9ac1538e05a0/core/java/android/view/Choreographer.java#L600-L608. Additionally (though probably less significant), any messages that were posted to the Looper (e.g. via View#post) since the last frame's CALLBACK_TRAVERSALS was called will be executed after the current frame's CALLBACK_TRAVERSALS.



([More info](https://w3c.github.io/requestidlecallback/))

Some notes to make up for my really poorly documented description at the top of this PR:
- This is a rough implementation that is intended to be experimental - the API shouldn't change (except for adding a timeout option) but the underlying behaviour can.
- Thanks for the details! This is my first time using the Choreographer API -- it looks like it makes more sense for us to put this after commit then, and that would be in agreement with diagram above. I couldn't find in the Choreographer source any way for us to add frame callback to `CALLBACK_COMMIT` -- is this what we would want to do or is there another way to do this?

> We really don't want to use the regular Choreographer here: ReactChoreographer allows us to be explicit about ordering of callback execution within the React framework, whereas if we use Choreographer directly, we might accidentally execute this before all of the ReactChoreographer callbacks.

Makes sense! I had it in my head that because I needed it to be outside of ReactChoreographer for reasons I can't quite remember because I originally wrote this about 2 months ago.

**Comment by astreet** [comment]
So, the important thing I wasn't really thinking about was that we are just concerned about idle time on the JS thread, not the UI thread. At least for now. Trying to estimate time remaining on the JS thread + time remaining on the UI thread for the (next) frame in which those resulting native UIManager module calls would be processed on the main thread is basically impossible.

So we can be satisfied for now by just invoking idle callbacks once we think we're done sending JS calls for the frame. These should be executed after all the `CallbackType.TIMERS_EVENTS` callbacks have executed (i.e. the `Events that make JS do things.`), so we should just add a new CallbackType in ReactChoreographer.

Note, however, that these Choreographer callbacks execute on the main thread. We may have already enqueued all the events and timers we need to call on the JS event queue, but we don't yet know at this time how long they'll take to execute. It could be no time, it could be multiple frames. This means that the idle handler in JS will need to actually check the amount of time remaining in the frame when it actually executes: that time could be zero or negative.

Finally, in the case where there are JS events/timers that take multiple frames to execute, we don't want to back up the JS event queue with a useless idle callback on each frame. To deal with this, I suggest we do what we do in EventDispatcher where we make sure to only have a single pending runnable in the queue at any point in time: https://github.com/facebook/react-native/blob/e42c6d4446dd1b5b9f68b1dfd2822c79036c2c3c/ReactAndroid/src/main/java/com/facebook/react/uimanager/events/EventDispatcher.java#L248-L254

**Comment by astreet** [comment]
I think the pattern for using this API should be:

```
// Taken from https://developers.google.com/web/updates/2015/08/using-requestidlecallback?hl=en
function myNonEssentialWork (deadline) {
  while (deadline.timeRemaining() > 0 && tasks.length > 0)
    doWorkIfNeeded();

  if (tasks.length > 0)
    requestIdleCallback(myNonEssentialWork);
}
```

Though experimental, a major part of this API will be figuring out how to ensure that the work you want to do in idle periods can be broken down into small/pauseable chunks. What may seem like ok chunk sizes on the emulator or on a phone with no load may bust the frame deadline on other phones.

**Comment by astreet** [comment]
I would also say that we should have only one idle callback actually registered natively if there are one or more idle callbacks registered in JS land, for simplicity. JS can handle turning that idle callback into multiple callbacks if necessary.

**Comment by hufeng** [comment]
cool

**Comment by brentvatne** [comment]
> I would also say that we should have only one idle callback actually registered natively if there are one or more idle callbacks registered in JS land, for simplicity. JS can handle turning that idle callback into multiple callbacks if necessary.

That's a great point!

Thanks for the insights @astreet, I'll try to improve this and push up changes over the next couple of weeks.

**Comment by mkonicek** [comment]
@brentvatne Is this ready for review now?

**Comment by facebook-github-bot-8** [comment]
@mkonicek would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

**Comment by ide** [comment]
This isn't yet ready for review, still need to address feedback.

**Comment by mkonicek** [comment]
👍 Thanks for the clarification @ide, I wasn't sure what the status was.

**Comment by facebook-github-bot-1** [comment]
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request we identified @mhorowitz as a potential reviewer. Could you take a look please or cc someone with more context?

**Comment by brentvatne** [comment]
Moved to #8569

**Pull request closed without merging**